QL: add unsigned_long type support

bpintea opened the pull request:
This PR introduces the `UNSIGNED_LONG` type to QL following its availability in ES (#60050).

The type is mapped to a BigInteger whose value is checked against the UL bounds.

The SQL will now support the type as literal and in the arithmetic functions; the non-arithmetic functions however are unchanged (i.e. they still require a long / int parameter where that is the case).

The type is version-gated: for the driver SQL clients (only) the server checks their version and in case this is lower than the one introducing the UL support, it fails the request, for queries, or simply hidden in catalog functions (similar to how `UNSUPPORTED` is currently treated in the similar case)

The JDBC tests are adjusted to read the (bwc) version of the driver they are run against and selectively disable part of the tests accordingly.

Closes #63312

bpintea committed:
Add unsigned_long type support to QL

This commit introduces the UNSIGNED_LONG type to QL, following it's
availabilty in ES. Just like in ES, the type is encoded as a BigInteger.

JDBC is now also UNSIGNED_LONG enabled.

In SQL, the type is exposed only to driver clients of a version equal or
greater to 7.11.0 (and all non-driver clients, irrespective of their
version).

bpintea committed:
style fixes

Remove unused imports.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/unsigned_long

bpintea committed:
Make JDBC QA tests version aware

Disable UNSIGNED_LONG testing for those bwc tests with drivers older
than 7.11 - first target release to support them.

bpintea committed:
Disable UL for 7.11 till after merging it in

Also, undo change of BigInteger.valueOf(randomNonNegativeLong()) to
randomBigInteger() for long-requiring tests.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/unsigned_long

bpintea committed:
Refactor UL gating based on version

Move the condition checking wheatehr unsigned_long fields should be
exposed to the clients from the IndexResolver to the Verifier,
SysColumns and SysTypes commands.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/unsigned_long

bpintea committed:
Minor clean-ups

Vars renaming, dead code removal.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/unsigned_long

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

astefan reviewed (commented):
Added a first set of reviews for almost half of the changes.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -87,9 +89,11 @@
  */
 public final class Verifier {
     private final Metrics metrics;
+    private final SqlConfiguration config;
```
If only the version is needed, it would be easier to pass it directly instead of passing the Config object.

bpintea replied:
I've changed it to only pass the Version.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysColumns.java:
```diff
@@ -169,7 +174,7 @@ private static void fillInRows(String clusterName, String indexName, Map<String,
 
             // skip the nested, object and unsupported types
-            if (isPrimitive(type)) {
+            if (isPrimitive(type) && (isDriver(mode) == false || isTypeSupportedInVersion(type, version))) {
```
Why is the drive mode important? What if the non-drivers want to be versioned (as is the plan for the rest APIs)?

bpintea replied:
> Why is the drive mode important?

The drivers would have issues with new types they don't know/understand. And that's why they send their version in each request.

We can relax this check to only consider the version, sure, but IMO we'd then need to understand why other clients would choose to be versioned and what guarantees we'd need to offer. Otherwise an unsigned long or decimal number for a REST client shouldn't pose an issue (since the JSON/CBOR/... library should understand that number).

bpintea replied:
The client check has been now dropped.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysTypes.java:
```diff
@@ -81,7 +83,9 @@ public SysTypes(Source source, int type) {
     @Override
     public final void execute(SqlSession session, ActionListener<Page> listener) {
-        Stream<DataType> values = SqlDataTypes.types().stream();
+        Stream<DataType> values = SqlDataTypes.types().stream()
```
Since this check occurred in a previous class, it should be extracted since otherwise its fairly obscure and repetitive.

bpintea replied:
Fixed.

costin commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/JdbcTestUtils.java:
```diff
@@ -5,25 +5,76 @@
 
+    static final Version JDBC_DRIVER_VERSION;
+
+    static {
```
A similar code is done inside the actual driver. Can we reuse it somehow so we don't have multiple mapping initializations that can (and eventually will) get out of sync? If SysTypes is versioned, does it make sense to initialize this map dynamically by calling the server API?
Or at least having a comment in *both* places so the code gets updated at once.

bpintea replied:
I've added comments.

Initialising a list with supported types is def. doable, but mapping those to Java classes would still require a "manual" map. Alternatively, we could extend `EsType` to include the class?

costin reviewed (changes requested):
First of, thanks for looking into this. It's a far reaching feature.
Second, it is a big PR making it hard not to produce but also review. I realize it's tricky however in the future please use a feature branch where the functionality can be built in stages (introducing the data type, checking the math operations, adding the type verification, adding the tests).
It's a tiny bit of overhead as it requires planning and waiting for reviews but it's much more digestable and easier to follow.

P.S. In fact since it took a while to get to it, I stopped at the test suite to give a first set of comments.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/type/SqlDataTypeConverter.java:
```diff
@@ -80,31 +81,25 @@ public static DataType commonType(DataType left, DataType right) {
             return DATETIME;
         }
         if (left == TIME) {
-            if (right == DATE) {
```
Why these removals?

bpintea replied:
They're not reachable: https://github.com/elastic/elasticsearch/pull/65145#discussion_r528549790

astefan commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/index/IndexResolverTests.java:
```diff
@@ -433,7 +433,7 @@ private void addFieldCaps(Map<String, Map<String, FieldCapabilities>> fieldCaps,
 
     private static IndexResolution mergedMappings(String indexPattern, String[] indexNames,
-            Map<String, Map<String, FieldCapabilities>> fieldCaps) {
+                                                  Map<String, Map<String, FieldCapabilities>> fieldCaps) {
```
Can you, please, revert this change?

bpintea replied:
Reverted.

astefan reviewed (changes requested):
1. I think you need to add a test in `FieldExtractorTestCase` as well.
2. I would look into adding a field type to `test_emp_copy` index and have some integration tests with this field type. Look at `DataLoader.loadEmpDatasetWithExtraIntoEs` where this index is created with some "extra" fields in it. There is a small section there where a `constant_keyword` and a `wildcard` field are added to the mapping. And further down in the method, there are values assigned to those fields. I think the easiest way of adding an `unsigned_long` is to put in it the salary value and then add some minimal tests to the csv-spec/sql-spec set. In `DataLoader` you'd do something like this:
```
            for (int f = 0; f < fields.size(); f++) {
                // an empty value in the csv file is treated as 'null', thus skipping it in the bulk request
                if (fields.get(f).trim().length() > 0) {
                    if (hadLastItem) {
                        bulk.append(",");
                    }
                    hadLastItem = true;
                    bulk.append('"').append(titles.get(f)).append("\":\"").append(fields.get(f)).append('"');
                    if (titles.get(f).equals("gender") && extraFields) {
                        bulk.append(",\"extra_gender\":\"Female\"");
                    } else if (titles.get(f).equals("salary") && extraFields) {
                        bulk.append(",\"ul\":\"" + fields.get(f) + "\"");
                    }
                }
```
 2.1. some ideas [here](https://github.com/elastic/elasticsearch/pull/58062)
 2.2. from the unsigned_long documentation I see a red-flag when it comes to using this field type in Scripts, as they seem they are unsupported. We need to decide if we add something to the documentation or handle this field type "nicely" and output a pretty error to the user in case the query they are building will end up using scripts (kind of hard to do I think). At the moment not even a simple `SELECT CAST(u_l AS INTEGER)` won't work for unsigned_long.
3. Do you think the tests we have so far answer the question "is running `SELECT unsigned_long FROM test` using an old JDBC driver version handled correctly both server and driver side?"

matriv commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/Check.java:
```diff
@@ -47,4 +49,13 @@ public static void isBoolean(Object obj) {
     }
+
+    // 18446744073709551615
+    public static final BigInteger UNSIGNED_LONG_MAX = BigInteger.ONE.shiftLeft(Long.SIZE).subtract(BigInteger.ONE);
```
Same here.

bpintea replied:
Fixed.

matriv reviewed (commented):
Left also some first comments, but need to check more.

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysTypesTests.java:
```diff
@@ -5,107 +5,148 @@
+                    List<String> types = new ArrayList<>();
+                    r.forEachRow(rv -> types.add((String) rv.column(0)));
+                    List<String> expectedTypes = ALL_TYPES.stream()
+                        .filter(t -> Mode.isDriver(mode) == false || isTypeSupportedInVersion(DataTypes.fromTypeName(t), version))
```
Another way of doing this is to:
1. have dedicated lists for base types +unsigned and check with and without.
2. check its presence only in versions w/ and w/o it.

bpintea replied:
Thanks, refactored.

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/SqlTestUtils.java:
```diff
@@ -56,12 +58,28 @@ public static SqlConfiguration randomConfiguration() {
 
     public static SqlConfiguration randomConfiguration(ZoneId providedZoneId) {
         return new SqlConfiguration(providedZoneId,
+            randomIntBetween(0,  1000),
```
Looks like bad formatting/identation.

bpintea replied:
C&P, fixed them all now.

costin reviewed (changes requested):
Left another round of comments.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/unsigned_long

bpintea committed:
Address reivew comments

- refactoring of the QA JDBC tests;
- introduce csv/sql-spec files;
- list of punctual changes as suggested.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feat/unsigned_long

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/JdbcTestUtils.java:
```diff
@@ -57,4 +108,9 @@ static long convertFromCalendarToUTC(long value, Calendar cal) {
     }
+
+    public static boolean isUnsignedLongSupported() {
+        // TODO: add equality only once actually ported to 7.11
+        return V_7_11_0.compareTo(JDBC_DRIVER_VERSION) < 0;
```
I've added a comment. Maybe a bit longer, but feel free to suggest an alternative, if finding it too long.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:
```diff
@@ -182,25 +185,16 @@ protected Object unwrapMultiValue(Object values) {
+                Number result = null;
+                try {
+                    // TODO: don't mapper modules expose _source parsing methods? should the _source be (re)validated?
+                    result = dataType == UNSIGNED_LONG ? new BigInteger(values.toString()) : numberType(dataType).parse(values, true);
```
Thanks. I'd had a look it, but found it heavier than what would be ideal here (we'd unnecessarily duplicate many of the checks with little net gain).

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/Arithmetics.java:
```diff
@@ -31,6 +34,22 @@ default Object wrap(Object l, Object r) {
         }
     }
 
+    private static BigInteger unsignedLongOperation(Number l, Number r, BiFunction<BigInteger, BigInteger, BigInteger> op) {
```
Moved.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/Check.java:
```diff
@@ -47,4 +49,13 @@ public static void isBoolean(Object obj) {
+    // 18446744073709551615
+    public static final BigInteger UNSIGNED_LONG_MAX = BigInteger.ONE.shiftLeft(Long.SIZE).subtract(BigInteger.ONE);
+
+    public static void isUnsignedLong(BigInteger bi) {
```
> There's a slightly different out of range check here

The signum is a pre-calculated value in BigInteger and readily available and so the comparison against `.ZERO` costlier.

> Also, I am not sure this is the right place for this kind of check.

Fair enough. I've taken Mario's [suggestion](https://github.com/elastic/elasticsearch/pull/65145#discussion_r528770855) and put it in a utility class. I do believe this check should have it's own function, rather than inlined, since it's the very definition of the "unsigned long" (i.e. an arbitrary range limitation of a BigInteger, in our case). This function could be placed in respective user classes too, but I'd personally avoid the duplication. Anyways, open for improvement suggestions.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/Arithmetics.java:
```diff
@@ -42,6 +61,10 @@ public static Number add(Number l, Number r) {
         if (l instanceof Float || r instanceof Float) {
             return Float.valueOf(l.floatValue() + r.floatValue());
         }
+        BigInteger bi = unsignedLongOperation(l, r, BigInteger::add);
```
Fixed, part of implementing Costin's [suggestions](https://github.com/elastic/elasticsearch/pull/65145#discussion_r528898593).

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/PreparedStatementTestCase.java:
```diff
@@ -61,29 +76,33 @@ public void testSupportedTypes() throws SQLException {
-                statement.setObject(18, utilDateVal);
-                statement.setObject(19, localDateTimeVal);
+                int index = 1;
+                statement.setString(index ++, stringVal);
```
Thanks, fixed.

bpintea commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/ExpressionBuilder.java:
```diff
@@ -221,15 +221,20 @@ public Literal visitIntegerLiteral(EqlBaseParser.IntegerLiteralContext ctx) {
+        if (value instanceof BigInteger) {
+            type = DataTypes.UNSIGNED_LONG;
+        } else {
+            assert value instanceof Long : "Expected value [" + value + "] of type Long but got: " + value.getClass();
```
I've removed the assertion.
(They seem to still be rightfully used throughout ES codebase though.)

bpintea commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java:
```diff
@@ -176,6 +176,7 @@ public void testAliasErrors() {
     // Test all elasticsearch numeric field types
     public void testNumeric() {
         final IndexResolution idxr = loadIndexResolution("mapping-numeric.json");
+        accept(idxr, "foo where unsigned_long_field == 0");
```
Moved it.
But generally, I was trying to enter the range progression at the right point (i.e. long -> int -> short etc.), rather then "append to the code".

bpintea commented on x-pack/plugin/eql/src/test/resources/mapping-numeric.json:
```diff
@@ -10,6 +10,9 @@
         "@timestamp" : {
             "type" : "date"
         },
+        "unsigned_long_field" : {
```
Moved it.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:
```diff
@@ -182,25 +185,16 @@ protected Object unwrapMultiValue(Object values) {
-                    return result;
+                Number result = null;
+                try {
+                    // TODO: don't mapper modules expose _source parsing methods? should the _source be (re)validated?
```
I've removed it altogether.
(The gist of it being about my expectation that mapper modules have a clear-cut function taking string - ie. from json - and outputting their particular object. This isn't the case, I've removed the comment.)

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:
```diff
@@ -182,25 +185,16 @@ protected Object unwrapMultiValue(Object values) {
+                Number result = null;
+                try {
+                    // TODO: don't mapper modules expose _source parsing methods? should the _source be (re)validated?
+                    result = dataType == UNSIGNED_LONG ? new BigInteger(values.toString()) : numberType(dataType).parse(values, true);
```
Extracted it.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/Arithmetics.java:
```diff
@@ -31,6 +34,22 @@ default Object wrap(Object l, Object r) {
         }
     }
 
+    private static BigInteger unsignedLongOperation(Number l, Number r, BiFunction<BigInteger, BigInteger, BigInteger> op) {
```
Thanks for catching that. Fixed/refactored it.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/Arithmetics.java:
```diff
@@ -140,10 +185,16 @@ static Number negate(Number n) {
         }
+        if (n instanceof BigInteger) {
+            if (((BigInteger) n).signum() != 0) {
+                throw new ArithmeticException("unsigned_long overflow");
```
Trying to negate a (positive) unsigned long results into a BitInteger out of unsigned_long's range. Just like casting a negative one to UL. So in the respect I would think it's correct.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DataTypeConverter.java:
```diff
@@ -334,15 +363,61 @@ public static int safeToInt(long x) {
     }
 
-    public static long safeToLong(double x) {
+    public static long safeDoubleToLong(double x) {
```
The original method didn't get renamed, strictly "speaking", but changed to take a Number. But also needed to add a new one (safeDoubleToLong) that takes doubles only. Since they both take doubles, I wanted to avoid the ambiguity.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DataTypeConverter.java:
```diff
@@ -334,15 +363,61 @@ public static int safeToInt(long x) {
+    }
+
+    // 18446744073709551615.0
+    private static final double UNSIGNED_LONG_MAX_AS_DOUBLE = UNSIGNED_LONG_MAX.doubleValue();
```
Fixed.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DataTypes.java:
```diff
@@ -102,6 +105,10 @@ public static DataType fromJava(Object value) {
             return LONG;
         }
+        if (value instanceof BigInteger) {
+            // TODO: range check needed at all? (((BigInteger) value).signum() < 0
```
> the output from ES should be okay

Great, thanks. (Comment removed.)

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/Check.java:
```diff
@@ -47,4 +49,13 @@ public static void isBoolean(Object obj) {
+    // 18446744073709551615
+    public static final BigInteger UNSIGNED_LONG_MAX = BigInteger.ONE.shiftLeft(Long.SIZE).subtract(BigInteger.ONE);
+
+    public static void isUnsignedLong(BigInteger bi) {
```
Function removed from that class.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/StringUtils.java:
```diff
@@ -326,6 +327,23 @@ public static long parseLong(String string) throws QlIllegalArgumentException {
         }
     }
 
+    public static Number parseInteger(String string) throws QlIllegalArgumentException {
```
I've changed it to `parseIntegral`.

FWIW, the function's called from `ExpressionBuilder#visitIntegerLiteral`. Additionally, I believe worth disambiguation is also DataType#isInteger()/isRational() (btw, rational numbers are a supertype of fixed point numbers).

bpintea commented on x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/BinaryArithmeticProcessorTests.java:
```diff
@@ -45,45 +48,129 @@ public void testAdd() {
         assertEquals(10, ba.process(null));
     }
 
+    public void testUnsignedLongAdd() {
```
Changed accordingly.

bpintea commented on x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/BinaryArithmeticProcessorTests.java:
```diff
@@ -45,45 +48,129 @@ public void testAdd() {
         assertEquals(4, ba.process(null));
     }
 
+    public void testUnsignedLongSub() {
```
Changed.

bpintea commented on x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/BinaryArithmeticProcessorTests.java:
```diff
@@ -45,45 +48,129 @@ public void testAdd() {
         assertEquals(21, ba.process(null));
     }
 
+    public void testUnsignedLongMul() {
```
Changed.

bpintea commented on x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/BinaryArithmeticProcessorTests.java:
```diff
@@ -45,45 +48,129 @@ public void testAdd() {
         assertEquals(2.33, ((Number) ba.process(null)).doubleValue(), 0.01d);
     }
 
+    public void testUnsignedLongDiv() {
```
Changed.

bpintea commented on x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/BinaryArithmeticProcessorTests.java:
```diff
@@ -45,45 +48,129 @@ public void testAdd() {
         assertEquals(1, ba.process(null));
     }
 
+    public void testUnsignedLongMod() {
```
Changed.

bpintea commented on x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/BinaryArithmeticProcessorTests.java:
```diff
@@ -45,45 +48,129 @@ public void testAdd() {
     }
-    
+
+    public void testUnsignedLongNegate() {
```
Changed.

bpintea commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcPreparedStatement.java:
```diff
@@ -111,7 +114,7 @@ public void setInt(int parameterIndex, int x) throws SQLException {
     @Override
     public void setLong(int parameterIndex, long x) throws SQLException {
-        setObject(parameterIndex, x, Types.BIGINT);
+        setObject(parameterIndex, x, BIGINT);
```
Style aligned.

bpintea commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/JdbcPreparedStatement.java:
```diff
@@ -204,7 +207,7 @@ public void setObject(int parameterIndex, Object x) throws SQLException {
         // {@code TypeConverter.fromJavaToJDBC(x.getClass())} will report the implementing class as not being supported.
         checkKnownUnsupportedTypes(x);
-        setObject(parameterIndex, x, TypeUtils.of(x.getClass()).getVendorTypeNumber(), 0);
+        setObject(parameterIndex, x, TypeUtils.of(x.getClass()).getVendorTypeNumber(), scaleOrLenght(x));
```
Fixed.

bpintea commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/TypeConverter.java:
```diff
@@ -342,6 +349,12 @@ private static Byte asByte(Object val, EsType columnType, String typeString) thr
             case INTEGER:
             case LONG:
                 return safeToByte(((Number) val).longValue());
+            case UNSIGNED_LONG:
```
I've extracted that into a method.
(My initial reason not too is to avoid the extra check, given that the switch jumps already at the right point.)
`byteValueExact()` will raise an exception if truncation would occur, which is what we want for the method to be "safe".

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -231,6 +235,18 @@ public Verifier(Metrics metrics) {
 
                 failures.addAll(localFailures);
             });
+
```
Extracted into `checkClientSupportsDataTypes`.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -231,6 +235,18 @@ public Verifier(Metrics metrics) {
                 failures.addAll(localFailures);
             });
+
+            if (config.version() != null) {
```
It's now always initialised to `CURRENT` in SqlConfiguration.
It used to be null if it hadn't been provided by the client (i.e. non-driver).

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -231,6 +235,18 @@ public Verifier(Metrics metrics) {
             });
+
+            if (config.version() != null) {
+                List<LogicalPlan> projects = plan.collectFirstChildren(x -> x instanceof Project);
```
I'd added a comment about it [here](https://github.com/elastic/elasticsearch/pull/65145#discussion_r528548116) and some tests for it.

We can do that, but just to make sure: the "top" project only will result in `unsigned_long`s being sent to the client and thus cause issues with old clients. I believe the intent is to cut the bwc at the API level, not features level.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -231,6 +235,18 @@ public Verifier(Metrics metrics) {
+
+            if (config.version() != null) {
+                List<LogicalPlan> projects = plan.collectFirstChildren(x -> x instanceof Project);
+                if (projects.size() > 0) { // not selecting just literals
```
I've removed the comment.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -231,6 +235,18 @@ public Verifier(Metrics metrics) {
+                if (projects.size() > 0) { // not selecting just literals
+                    ((Project) projects.get(0)).projections().forEach(e -> {
+                        if (e.resolved() && isTypeSupportedInVersion(e.dataType(), config.version()) == false) {
+                            failures.add(fail(e, "Cannot use field [" + e.name() + "] with type [" + e.dataType() + "] unsupported " +
```
Added clarification.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/session/SqlConfiguration.java:
```diff
@@ -9,25 +9,28 @@
     private final TimeValue pageTimeout;
     private final Mode mode;
     private final String clientId;
+    @Nullable
```
Thanks, fixed that.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/session/Compatibility.java:
```diff
@@ -0,0 +1,39 @@
+import static org.elasticsearch.Version.V_7_11_0;
+import static org.elasticsearch.xpack.ql.type.DataTypes.UNSIGNED_LONG;
+
+public final class Compatibility {
```
I went with `VersionCompabilityChecks`.
And yes, a follow-up PR should generify the checks.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/StringUtils.java:
```diff
@@ -326,6 +327,23 @@ public static long parseLong(String string) throws QlIllegalArgumentException {
+            } catch (ArithmeticException ae) {
+                throw new QlIllegalArgumentException("Number [{}] is too large", string);
+            }
+        } catch (NumberFormatException ex) {
```
I've reduce the `try` block.

bpintea commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/TypeUtils.java:
```diff
@@ -39,6 +41,7 @@ private TypeUtils() {}
             EsType.SHORT, EsType.INTEGER, EsType.LONG,
             EsType.FLOAT, EsType.HALF_FLOAT, EsType.SCALED_FLOAT, EsType.DOUBLE, EsType.DATETIME);
 
+    public static final int LONG_MAX_LENGTH = String.valueOf(Long.MAX_VALUE).length(); /* type length value as defined in ES  */
```
Fixed it.

bpintea commented on x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/TypeConverterTests.java:
```diff
@@ -61,15 +89,25 @@ public void testDateAsNative() throws Exception {
     }
+
+    private <T> Object convertAsNative(Object value, Class<T> nativeType) throws Exception {
+        Object copy = throughXContent(value);
```
Thanks, fixed (c&p err).

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/PreparedStatementTestCase.java:
```diff
@@ -27,11 +29,23 @@
 public abstract class PreparedStatementTestCase extends JdbcIntegrationTestCase {
 
+    static final BigInteger UNSIGNED_LONG_MAX = BigInteger.ONE.shiftLeft(Long.SIZE).subtract(BigInteger.ONE);
+    EsType UNSIGNED_LONG = isUnsignedLongSupported() ? EsType.valueOf("UNSIGNED_LONG") : null;
```
This bit was reworked.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:
```diff
@@ -182,25 +185,16 @@ protected Object unwrapMultiValue(Object values) {
-                    return result;
+                Number result = null;
+                try {
+                    // TODO: don't mapper modules expose _source parsing methods? should the _source be (re)validated?
```
I've removed it altogether.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/arithmetic/Arithmetics.java:
```diff
@@ -140,10 +185,16 @@ static Number negate(Number n) {
             return Float.valueOf(-n.floatValue());
         }
+        if (n instanceof BigInteger) {
+            if (((BigInteger) n).signum() != 0) {
```
To this point a BigInteger can only be `>= 0`. So the check is there for the `> 0` part, since negating a non-negative unsigned long should trigger an error response. I.e. `!= 0`.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DataTypeConverter.java:
```diff
@@ -334,15 +363,61 @@ public static int safeToInt(long x) {
+    }
+
+    // 18446744073709551615.0
+    private static final double UNSIGNED_LONG_MAX_AS_DOUBLE = UNSIGNED_LONG_MAX.doubleValue();
```
Sure, added it to a utility class.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DataTypeConverter.java:
```diff
@@ -72,6 +76,9 @@ public static DataType commonType(DataType left, DataType right) {
             if (left.isInteger()) {
                 // promote the highest int
                 if (right.isInteger()) {
+                    if (left == UNSIGNED_LONG || right == UNSIGNED_LONG) {
```
> if we have: -10::int x 1000::unsigned_long then we end up with a negative unsigned long which will throw an Exception.

Right, that should be the outcome. 
I would think the check is correct: if either numbers is a UL, the result should be a UL or an error if that's not possible.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/DataTypes.java:
```diff
@@ -102,6 +105,10 @@ public static DataType fromJava(Object value) {
             return LONG;
         }
+        if (value instanceof BigInteger) {
+            // TODO: range check needed at all? (((BigInteger) value).signum() < 0
```
Thanks for confirming. Removed TODO.

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/PreparedStatementTestCase.java:
```diff
@@ -42,6 +56,7 @@ public void testSupportedTypes() throws SQLException {
         byte byteVal = randomByte();
         short shortVal = randomShort();
         BigDecimal bigDecimalVal = BigDecimal.valueOf(randomDouble());
+        BigInteger bigIntegerVal = bigDecimalVal.abs().remainder(new BigDecimal(UNSIGNED_LONG_MAX)).toBigInteger();
```
This got refactored.

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/PreparedStatementTestCase.java:
```diff
@@ -61,29 +76,33 @@ public void testSupportedTypes() throws SQLException {
+                statement.setObject(index ++, calendarVal);
+                statement.setObject(index ++, utilDateVal);
+                statement.setObject(index ++, localDateTimeVal);
+                if (isUnsignedLongSupported()) {
```
Refactored.

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/PreparedStatementTestCase.java:
```diff
@@ -341,32 +365,42 @@ public void testSingleParameterMultipleTypes() throws SQLException {
                 statement.setShort(1, shortVal);
-                assertEquals(new Tuple<>(JDBCType.SMALLINT.getVendorTypeNumber(), shortVal), execute(statement));
+                assertEquals(new Tuple<>(SHORT.getName(), shortVal), execute(statement));
+
```
Refactored.

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetMetaDataTestCase.java:
```diff
@@ -14,19 +14,32 @@
+        }
+    }
 
     public void testValidGetObjectCalls() throws IOException, SQLException {
```
Refactored.

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetMetaDataTestCase.java:
```diff
@@ -36,30 +49,30 @@ public void testValidGetObjectCalls() throws IOException, SQLException {
-            + "test_keyword AS k, test_boolean AS bool, test_date AS dt FROM test";
-        doWithQuery(q, r -> assertColumnNamesAndLabels(r.getMetaData(), new String[] { "b", "i", "l", "s", "d", "f", "k", "bool", "dt" }));
+
+        q = "SELECT " + fieldsNames.stream().map(x -> x + " AS " + x.replace("_", "")).collect(Collectors.joining(", ")) + " FROM test";
```
Extracted.

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetTestCase.java:
```diff
@@ -64,23 +64,36 @@
+    static final EsType UNSIGNED_LONG;
+
     static {
+        fieldsNames.addAll(List.of(
```
Refactored.

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetTestCase.java:
```diff
@@ -89,6 +102,9 @@
         dateTimeTestingFields.put(new Tuple<>("test_double", 1d), EsType.DOUBLE);
         dateTimeTestingFields.put(new Tuple<>("test_float", 1f), EsType.FLOAT);
         dateTimeTestingFields.put(new Tuple<>("test_keyword", "true"), EsType.KEYWORD);
+        if (isUnsignedLongSupported()) {
```
Refactored.

[Offtopic remark]:
IMO this dedicated-methods refactoring enlarges the tests code base with little bottom line clarity. See `testGettingValidShortWithCasting` for instance -- anyone reading it would expect "short with casting" case to be covered there entirely -- that's what the method says. But then it's not: unsigned long is a special case, treated somewhere else (i.e. in the new `testGettingValidNumbersWithCastingFromUnsignedLong`). This might be clear "enough" now, but sometimes towards the end of 8.x dev cycle it'll likely no longer be the case.

And this could IMO be addressed by running both the driver and the QA tests of the same branch, rather than the mixed approach used currently.

bpintea commented on x-pack/plugin/sql/sql-proto/src/main/java/org/elasticsearch/xpack/sql/proto/SqlVersion.java:
```diff
@@ -23,6 +23,7 @@
     public final byte major;
     public final byte minor;
     public final byte revision;
+    public final byte build; // the build is required for id-compatibility with ES's version
```
This is no longer strictly required now TBH.

In a previous version of the PR I had Version turned into an SqlVersion and then back to Version, needed for testing purposes. This resulted in non-equal Version objects, because of this `build` value. The tests no longer require that, but it's still safer to have it like this, IMO.

bpintea commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysColumnsTests.java:
```diff
@@ -52,363 +58,267 @@
+        assertEquals(typeClass, sqlDataType(row).getClass());
+        assertEquals(typeClass, sqlDataTypeSub(row).getClass());
+
+        if (unsignedLongSupported) {
```
Refactored.

bpintea commented:
> * I think you need to add a test in `FieldExtractorTestCase` as well.

I've added csv- & sql-spec tests too.

> At the moment not even a simple SELECT CAST(u_l AS INTEGER) won't work for unsigned_long.

That should work (and there are tests for it too). But indeed, scripts (i.e. `GROUP BY some_function(unsigned_long)`) have issues. There's work going on - #64361 - but till then we can either leave [the exception](https://github.com/elastic/elasticsearch/blob/be86dd5625e805078de56cd1332fa052c569d71e/x-pack/plugin/mapper-unsigned-long/src/main/java/org/elasticsearch/xpack/unsignedlong/UnsignedLongLeafFieldData.java#L77) surface or replace it in the Verifier? (I'll look into the latter.)

> 3. Do you think the tests we have so far answer the question "is running SELECT unsigned_long FROM test using an old JDBC driver version handled correctly both server and driver side?"

Yes-ish: the bwc tests can only [go back as far as 7.9.0](https://github.com/elastic/elasticsearch/blob/be86dd5625e805078de56cd1332fa052c569d71e/x-pack/plugin/sql/qa/jdbc/build.gradle#L76). jar hell ensues if trying to go back further (I tried). I guess (?) we could try to port changes further back till 7.7.0, but: as far as unsigned_long is concerned, if it works for 7.9.0 it _should_ for 7.7.0 as well.
(FWIW I'll expand ODBC to run bwc tests too and starting from 7.7.0, but those aren't as exhaustive as JDBC's.)

bpintea committed:
Remove stale comment

- remoce comment
- undo primitive -> object change in one test.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/unsigned_long

bpintea committed:
Filter out UL from * expansion. Drop client check

- Filters out UL from star expansion (SELECT * ...). This is required
  since the verifier will otherwise later fail an otherwise
  "supportable" query for old clients.

- Drop checking for the client type when enforcing version
  compatibility. If the client inserts a version, we should answer
  according to that version's compability no matter the client.

bpintea committed:
Revert FieldExtractorTestCase: no float to UL

The UL mapper won't accept any floating point representation.

bpintea commented:
Thanks for the prompt (and generous!) feedback. I hope I've addressed all the raised points.
The UL support in scripts remaining a pending issue, though, that will need to be addressed in parallel, but hope to have this PR ready before then.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:
```diff
@@ -435,10 +435,16 @@ else if (plan instanceof OrderBy) {
             }
 
-            return result;
+            return filterUnsupportedProjections(result);
```
Why this change?

bpintea replied:
This is needed since, without it, if an index has an `unsigned_long` field and a pre-unsigned_long client (ex 7.7.0) executes a `SELECT * FROM ...` the "expanded" set of fields will contain this UL field that will get caught in the verifier (`#checkClientSupportsDataTypes()`), which checks if a query from an old client tries to select "new" fields that the client would not understand.

astefan commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:
```diff
@@ -435,10 +435,16 @@ else if (plan instanceof OrderBy) {
-        static List<NamedExpression> expandStar(UnresolvedStar us, List<Attribute> output) {
+        private List<NamedExpression> filterUnsupportedProjections(List<NamedExpression> projections) {
+            return projections.stream()
+                .filter(p -> p.resolved() == false || isTypeSupportedInVersion(p.dataType(), configuration.version()))
```
Why `||`? I would have thought `&&` is the one...

bpintea replied:
I think that if the fields is unresolved the type might not be available.

astefan reviewed (commented):
Left few comments.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java:
```diff
@@ -214,6 +213,10 @@ private static NumberType numberType(DataType dataType) {
         return NumberType.valueOf(dataType.esType().toUpperCase(Locale.ROOT));
     }
 
+    private static Number createNumeric(DataType dataType, Object value) {
```
Added on review comment: https://github.com/elastic/elasticsearch/pull/65145#discussion_r528889315
(Sorry for the long PR, I realise it's difficult to follow who said what about which piece, as a reviewer.)

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/StringUtils.java:
```diff
@@ -327,21 +327,20 @@ public static long parseLong(String string) throws QlIllegalArgumentException {
         } catch (NumberFormatException ex) {
             throw new QlIllegalArgumentException("Cannot parse number [{}]", string);
         }
+        if (bi.signum() < 0 || bi.compareTo(BigInteger.valueOf(Long.MAX_VALUE)) <= 0) {
```
> I may be missing something, but why if the sign of the bigInteger is "minus", we are still converting it?

We're converting it to long, in this case.

`parseIntegral()` - called from `ExpressionBuilder` - takes a string as input and tries to resolve it to a long, if possible - i.e. in case the number is negative; or positive, but within long's range - or an unsigned long if not. But since the BigInteger parsing accepts negative numbers, we need to check that explicitly still.

I guess the method could be refactored to only check once if the number is negative, but not sure it would make a relevant positive legibility difference overall.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/UnsignedLongUtils.java:
```diff
@@ -0,0 +1,36 @@
+
+import java.math.BigInteger;
+
+public abstract class UnsignedLongUtils {
```
Instead of using a dedicated utils just for unsigned longs, it would make sense to create a util class for number/numerics. So methods spread around such as  DataTypeConver#safeTo could be moved here
This can be a follow-up PR however for now I would rename this to NumericUtils.

bpintea replied:
Makes sense, changed.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -882,4 +889,18 @@ private static void checkCastOnInexact(LogicalPlan p, Set<Failure> localFailures
         }, Cast.class)), Filter.class);
     }
+
+    private static void checkClientSupportsDataTypes(LogicalPlan p, Set<Failure> localFailures, SqlVersion version) {
```
I think it's better to just take a look at the output of the plan, regardless of whether a select is used or not. That is you don't have to analyze the structure only its output.

bpintea replied:
Thanks, changed.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:
```diff
@@ -434,10 +435,16 @@ else if (plan instanceof OrderBy) {
             }
 
-            return result;
+            return filterUnsupportedProjections(result);
```
This part needs discussing. Hiding fields that exist from the resolution can easily become problematic.
A. SELECT * FROM table -> returns only supported results
B. SELECT unsupported_field FROM (SELECT * FROM TABLE) -> returns unknown field despite the field existing and knowing about it due to filtering

Who should win? A or B? I like the friendliness of A however I would argue that B should be the behaviour. That is if the users asks all fields, including those unsupported, we should do that. Otherwise we'll end with incorrect view of the worlds.
The same approach is used for dealing with mapping conflicts - we are aware of them however as long as the field is not used, things are good. Doing SELECT * on them though will bubble the exception.
That's because we don't want to resolve user conflicts, rather just be lazy about them.

bpintea replied:
So, not positive I understand, sorry if getting it wrong, but:

> That is if the users asks all fields, including those unsupported, we should do that. Otherwise we'll end with incorrect view of the worlds.

We could try. (That's not what we do with ranges (or nested), though, for instance. It's simply unsupported/hidden.)
But the reason to hide it here, however, is:  what should an "old" driver do with an unsigned_long, even if the user explicitly asks for it. It'll still fail, but instead of doing it with an explicit server-generated error message, it'll fail with a more cryptic driver error.

We could maybe be more liberal if we (re)introduce restricting types based on client type (i.e. `isDriver()`)?

Additionally, that some apps do automatically run "SELECT *" (I'll share a log later).

If not a driver, the client could potentially choose (a) not to be versioned - i.e. skip communicating the version and so asking for server's version behaviour - or (b) choose a "recent enough" one.

But just to be sure:
> B. SELECT unsupported_field FROM (SELECT * FROM TABLE) -> returns unknown field despite the field existing and knowing about it due to filtering

This currently would still fail, assuming that `unsupported_field ` means in this case an `unsigned_long` field.

costin replied:
I don't see an agreed conclusion to this discussion - please put it on the agenda for the next meeting.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -231,6 +236,8 @@ public Verifier(Metrics metrics) {
                 failures.addAll(localFailures);
             });
+
+            checkClientSupportsDataTypes(plan, failures, version);
```
This check is useful only when NOT doing SELECT * - that is when the consumer specifies each field individually. The two approaches need to be aligned - either we filter all the way or not at all.

bpintea replied:
In a version-gated answer, the `unsigned_long` fields would appear nowhere in SQL catalog results. So `SELECT *` expansion is consistent with this check. If the user "discovers" the fields "out-of-band" though and explicitly inserts an unsigned_long field in the query, that'll be caught by this check.
Again, I think we do this for other "unsupported" field types too (just not version-gated).

To make sure: I think aligning is great, just not sure on how to improve this, specifically.

costin replied:
Similar to the above - worth discussing.
My take is to currently filter things for * but return unsigned long as UNSUPPORTED for old clients.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysColumns.java:
```diff
@@ -169,7 +173,7 @@ private static void fillInRows(String clusterName, String indexName, Map<String,
 
             // skip the nested, object and unsupported types
-            if (isPrimitive(type)) {
+            if (isPrimitive(type) && isTypeSupportedInVersion(type, version)) {
```
I believe this is incorrect - any column using of an unsupported type should be exposed as such - unsupported, not hidden from view.

astefan replied:
@costin `isPrimitive` is already eliminating the `UNSUPPORTED` types from `sys columns`. If `unsigned_long` would have been configured as `UNSUPPORTED` instead, it wouldn't have been part of `sys columns` output by default.
`SHOW COLUMNS` on the other hand should display `unsupported` for an old driver version and `unsigned_long` for a newer driver version. Which, I don't think it happens atm.

bpintea replied:
Right. And we do that to hide these fields from "nosy" apps, that would otherwise still SELECT them (since `UNSUPPORTED` is just a vendor-specific type, like the rest of the types).

costin reviewed (changes requested):
Left another round of comments

astefan reviewed (commented):
I've left some minor comments with one exception: `SHOW COLUMNS` should behave differently, for consistency, when it comes to different driver versions?
At the moment, I _don't think_ we do anything when it comes to this command for different driver versions when, imo, we should return `unsupported` for a driver that doesn't know anything about `unsigned_long` types and return `unsigned_long` for a driver that does know about it.

bpintea committed:
Address review comments

- rename class;
- make one function private;
- undo one line break.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/unsigned_long

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetMetaDataTestCase.java:
```diff
@@ -26,40 +35,73 @@
         });
+    }
+
+    public void doTestValidGetObjectCalls(List<String> fieldsNames) throws IOException, SQLException {
```
"Private-ized" it, moved it below.

bpintea commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/FieldAttributeTests.java:
```diff
@@ -278,6 +285,92 @@ public void testGroupByAmbiguity() {
+
+            for (SqlVersion v : List.of(introducingUnsignedLong, postUnsignedLong)) {
+                analyzer = new Analyzer(SqlTestUtils.randomConfiguration(v), functionRegistry, getIndexResult,
+                    verifier);
```
Aligned.

bpintea commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/FieldAttributeTests.java:
```diff
@@ -278,6 +285,92 @@ public void testGroupByAmbiguity() {
+        String queryWithArithmetic = "SELECT unsigned_long + 1 AS unsigned_long FROM test";
+        String queryWithCast = "SELECT long + 1::unsigned_long AS unsigned_long FROM test";
+
+        SqlVersion introducingUnsignedLong = INTRODUCING_UNSIGNED_LONG;
```
Removed it, thanks (left-over from a refactoring).

bpintea commented:
> SHOW COLUMNS should behave differently, for consistency, when it comes to different driver versions?

I tend to think that yes, if the user specified an old ("enough") version, we should filter `unsigned_long`s out.
OTOH, I believe the `SHOW`s are pure human-specific commands and a user would not normally issue them  when connecting through a driver. If connecting through the CLI or REST, `unsigned_long`s would not be an issue.

Anyways, unless anyone raises a concern, I'll have `SHOW` be version-aware.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/unsigned_long

bpintea committed:
Add 'SHOW COLUMNS' tests

Add tests for `SHOW COLUMNS`.

bpintea committed:
Switch UNSIGNED_LONG to NUMERIC xDBC type

Most client applications don't support/check unsigned types.
Furthermore, seeing a BIGINT type, they assume it's signed (although
xDBC APIs allow checking). Consequently the NUMERIC type offers more conversion
safty.

bpintea committed:
Update test

- update SHOW commands to newly changed NUMERIC type;
- pass a BigInteger value that won't fit into a Long for field extr.
  test.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feat/unsigned_long

bpintea committed:
Style fix

Update one comment line.

bpintea commented:
> Anyways, unless anyone raises a concern, I'll have SHOW be version-aware.

This test has been now added.

Additionally, after some more testing, it's obvious that most client applications don't support (or check for) unsigned types.
Advertising `UNSIGNED_LONG` as a `BIGINT` type makes them assume it's signed (although
xDBC APIs allow checking). Consequently I've changed the xDBC type to NUMERIC, which offer more conversion safety.

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/unsigned_long

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/unsigned_long

bpintea committed:
Merge remote-tracking branch 'upstream/master' into feat/unsigned_long

bpintea committed:
License header fix

Fix headers of new files added to this PR before EL2.

bpintea committed:
Merge branch 'master' into feat/unsigned_long

bpintea committed:
Update target release.Update test to 4-dig. years

- Update target release to 7.15
- Update test to the newly restricted 4-digits year.

bpintea committed:
Merge branch 'master' into feat/unsigned_long

bpintea committed:
Merge branch 'master' into feat/unsigned_long

bpintea committed:
Merge branch 'master' into feat/unsigned_long

bpintea committed:
Add Painless-driven testing

Add queries to test unsigned long going through Painless scripts.

bpintea committed:
Merge branch 'master' into feat/unsigned_long

bpintea committed:
Merge branch 'master' into feat/unsigned_long

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feat/unsigned_long

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:
```diff
@@ -92,13 +93,13 @@
     private final Verifier verifier;
 
-    public Analyzer(Configuration configuration, FunctionRegistry functionRegistry, IndexResolution results, Verifier verifier) {
+    public Analyzer(SqlConfiguration configuration, FunctionRegistry functionRegistry, IndexResolution results, Verifier verifier) {
```
Worth looking whether the version option needs to be moved into the configuration base class.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/plan/logical/command/ShowColumns.java:
```diff
@@ -82,23 +84,23 @@ public void execute(SqlSession session, ActionListener<Page> listener) {
             DataType dt = field.getDataType();
             String name = e.getKey();
-            if (dt != null) {
+            if (dt != null && isTypeSupportedInVersion(dt, version)) {
```
Same question here: if we show unsupported fields, we should do the same for supported types when reporting to old clients.

bpintea replied:
Removed the filtering in `SHOW COLUMNS`. Note that this is still kept in `SYS COLUMNS`, to prevent (some) clients from automatically selecting these columns.

costin reviewed (commented):
Left a few comments.
I believe the jdbc enum breaks bwc; my remarks around filtering types likely just need some refreshing.

Luegg commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java:
```diff
@@ -729,21 +730,27 @@ public Literal visitDecimalLiteral(DecimalLiteralContext ctx) {
+        Number value;
         try {
-            value = Long.valueOf(StringUtils.parseLong(tuple.v2()));
+            value = StringUtils.parseIntegral(tuple.v2());
```
Could `parseIntegral` also directly return the best fitting `Number` type? Right now the downsizing from `BigInteger` to `Long` happens in `parseIntegral` and from `Long` to `Integer` here and in  the EQL `ExpressionBuilder`

bpintea replied:
Good point, fixed, thanks.

Luegg commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java:
```diff
@@ -729,21 +730,27 @@ public Literal visitDecimalLiteral(DecimalLiteralContext ctx) {
+        if (value instanceof BigInteger) {
+            type = DataTypes.UNSIGNED_LONG;
+        } else {
+            assert value instanceof Long : "Expected value [" + value + "] of type Long but got: " + value.getClass();
```
Assert should probably be replaced with an Exception?

bpintea replied:
Refactored away.

Luegg reviewed (commented):
Apart from the other comments I think there is also a problem with scripting in this PR:

Given a mapping
```
{
  "mappings": {
    "properties": {
      "l": { 
        "type": "unsigned_long"
      }
    }
  }
}
```

and a doc
```
{
  "l" : "18446744073709551614"
}
```

The query `select l::string from ulong group by 1` produces `-2` but expected would be a string `18446744073709551614`.

astefan commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/ShowFunctionsTests.java:
```diff
@@ -30,4 +58,66 @@ public void testShowFunctions() throws Exception {
+
+        for (SqlVersion version : versions) {
+            List<List<?>> rows = new ArrayList<>();
+            ShowColumns.fillInRows(loadMapping("mapping-multi-field-variation.json", true), null, version, rows);
```
Why do you loadMapping every time? I see this as unnecessary time-consuming operations.

bpintea replied:
The `ShowColumns#fillInRows()` behaves version-dependent.

bpintea replied:
Note: these two "SHOW COLUMNS" tests have now been moved to dedicated class.

bpintea replied:
Thanks @astefan for the clarification, now fixed this issue with looped mapping reading as well.

astefan commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysColumnsTests.java:
```diff
@@ -130,6 +138,35 @@ public void testInNonDriverMode() {
     }
 
+    public void testUnsignedLongFiltering() {
+        List<SqlVersion> versions = List.of(
+            SqlVersion.fromId(INTRODUCING_UNSIGNED_LONG.id - SqlVersion.MINOR_MULTIPLIER),
+            INTRODUCING_UNSIGNED_LONG,
+            SqlVersion.fromId(INTRODUCING_UNSIGNED_LONG.id + SqlVersion.MINOR_MULTIPLIER),
+            SqlVersion.fromId(Version.CURRENT.id)
+        );
```
This list is the same as https://github.com/elastic/elasticsearch/pull/65145/files#diff-8f9e4dcb53d1a3bb501882f5c67dfb19be7a0606fd27c2762060721d8d7fe2f3R108. Worth extracting it to a common place?

bpintea replied:
Makes sense, extracted and reused.

astefan commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysColumnsTests.java:
```diff
@@ -130,6 +138,35 @@ public void testInNonDriverMode() {
+                SysColumns.fillInRows(
+                    "test",
+                    "index",
+                    loadMapping("mapping-multi-field-variation.json", true),
```
Same here about loading the same mapping file multiple times. Isn't it enough to load the file once?

bpintea replied:
Also `SysColumns#fillInRows()` behaves version-dependent.

bpintea replied:
Fixed.

astefan commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysTablesTests.java:
```diff
@@ -353,7 +353,7 @@ private SqlTypedParamValue param(Object value) {
             new FunctionRegistry(),
             IndexResolution.valid(test),
-            new Verifier(new Metrics())
+            new Verifier(new Metrics(), SqlTestUtils.TEST_CFG.version())
```
All these tiny changes to tests classes to introduce the version in the constructor shouldn't, hopefully, be needed anymore.

bpintea replied:
Indeed, reverted these changes.

astefan commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/plan/logical/command/sys/SysTypesTests.java:
```diff
@@ -132,6 +139,33 @@ public void testSysTypes() {
+    public void testUnsignedLongFiltering() {
+        Set<SqlVersion> versions = new HashSet<>(
+            List.of(
+                SqlVersion.fromId(INTRODUCING_UNSIGNED_LONG.id - SqlVersion.MINOR_MULTIPLIER),
+                INTRODUCING_UNSIGNED_LONG,
+                SqlVersion.fromId(INTRODUCING_UNSIGNED_LONG.id + SqlVersion.MINOR_MULTIPLIER),
+                SqlVersion.fromId(Version.CURRENT.id)
```
Same here: this list can be extracted in a common place and re-used.

bpintea replied:
Suggestion applied.

astefan reviewed (commented):
Impressive amount of changes. In a good way.
There are some aspects that can be improved. Left some comments.

bpintea committed:
Merge branch 'master' into feat/unsigned_long

bpintea committed:
Extend value extraction as unsigned long

The values returned for unsigned long fields needed further value and
type conversion, like in the case of scripts returns, or type promotion,
like in the case of the extractors.

Testing has also been extended to include unsigned long values exceeding
the Long range all the way up unsigned long upper limit. This revealed
the need to further change some aggs implementation, like MAX/MIN
delegating to LAST/FIRST.

bpintea committed:
Address review comments

- extend comments;
- move client's version argument from Verifier's constructor to methods;
- extract unsigned long testing version to reusable var;
- StringUtils#parseIntegral() now returns the "smallest" fitting Number;
- adjust some member and methods' visibility.

bpintea committed:
Merge branch 'master' into feat/unsigned_long

bpintea commented:
@elasticmachine run elasticsearch-ci/part-2

bpintea commented:
> I think there is also a problem with scripting in this PR

@Luegg, there was indeed, thanks. Added better test data and hopefully ironed out all issues.

bpintea commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/type/SqlDataTypeConverterTests.java:
```diff
@@ -513,6 +535,30 @@ public void testConversionToBoolean() {
         }
     }
 
+    public void testConversiontoUnsignedLong() {
```
Thanks, fixed. (There was one more instance of this typo.)

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -1008,6 +1015,27 @@ private static void checkCastOnInexact(LogicalPlan p, Set<Failure> localFailures
+                            + "] unsupported in version ["
+                            + version
+                            + "], upgrade required (to version ["
+                            + INTRODUCING_UNSIGNED_LONG
```
Good point. I've added a new method to return the introducing version of a type.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -93,9 +96,11 @@
-    public Verifier(Metrics metrics) {
+    public Verifier(Metrics metrics, SqlVersion version) {
         this.metrics = metrics;
+        this.version = version;
```
This got refactored away.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/aggregate/Sum.java:
```diff
@@ -38,7 +39,8 @@ public Sum replaceChildren(List<Expression> newChildren) {
     public DataType dataType() {
-        return field().dataType().isInteger() ? LONG : DOUBLE;
+        DataType dt = field().dataType();
+        return dt.isInteger() ? (dt == UNSIGNED_LONG ? UNSIGNED_LONG : LONG) : DOUBLE;
```
Yes, I've eventually changed the type. Aggs return double for other integral types too, but we change the type where possible (as per SQL's practice, also spec'd in our [docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-functions-aggs.html#sql-functions-aggs-sum)). But in this case, the higher end of the unsigned_long the double won't be precise enough to recover the integral value, so we'll need to return double indeed.

Laterally, MIN/MAX aggs on unsigned_longs are also affected by the double-return, so they now delegate to FIRST/LAST.

costin replied:
Let's double check this with the ES (agg) team - we should have as little workarounds for this as possible. If the situation is unclear/unsupported then this should be signaled (either by us or ideally ES).

bpintea commented on x-pack/plugin/sql/jdbc/src/test/java/org/elasticsearch/xpack/sql/jdbc/JdbcPreparedStatementTests.java:
```diff
@@ -234,6 +245,51 @@ public void testSettingLongValues() throws SQLException {
+        assertEquals(bi.doubleValue(), value(jps));
+        assertEquals(DOUBLE, jdbcType(jps));
+
+        jps.setObject(1, bi, Types.FLOAT);
```
Good point, added.
(I guess we should probably extend the other existing tests with REAL.)

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/JdbcTestUtils.java:
```diff
@@ -7,29 +7,43 @@
     static final LocalDate EPOCH = LocalDate.of(1970, 1, 1);
 
+    static final String UNSIGNED_LONG_TYPE_NAME = "UNSIGNED_LONG";
+    public static final BigInteger UNSIGNED_LONG_MAX = BigInteger.ONE.shiftLeft(Long.SIZE).subtract(BigInteger.ONE);
```
Fixed.

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/ResultSetTestCase.java:
```diff
@@ -2199,23 +2431,45 @@ private void createTestDataForBooleanValueTests() throws IOException {
         });
     }
 
+    protected static void indexTestFieldsDoc(String docId, Object... values) throws IOException {
```
Fixed.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/PlanExecutor.java:
```diff
@@ -58,13 +57,22 @@ public PlanExecutor(Client client, IndexResolver indexResolver, NamedWriteableRe
+            functionRegistry,
+            indexResolver,
+            preAnalyzer,
+            new Verifier(metrics, cfg.version()),
```
Good point, thanks, fixed.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/parser/ExpressionBuilder.java:
```diff
@@ -729,21 +730,27 @@ public Literal visitDecimalLiteral(DecimalLiteralContext ctx) {
+        if (value instanceof BigInteger) {
+            type = DataTypes.UNSIGNED_LONG;
+        } else {
+            assert value instanceof Long : "Expected value [" + value + "] of type Long but got: " + value.getClass();
```
This had been refactored away.

bpintea commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/EsType.java:
```diff
@@ -19,6 +19,7 @@
     SHORT(Types.SMALLINT),
     INTEGER(Types.INTEGER),
     LONG(Types.BIGINT),
+    UNSIGNED_LONG(Types.NUMERIC),
```
> old clients sending EsType of double will get interpreted as UNSIGNED_LONG

Can of course change it, but to double check: JDBC's handling of SQL types is string-, not id-based (i.e. `"type": "DOUBLE"`, not `"type": 7`) and no bwc test revealed any issue so far. I believe the types in the JDBC aren't serialised either. The tests additionally make use of class-to-EsType resolution (but not ordinal to type).
But will change if this is indeed an issue.

costin replied:
Since the EsType enum is public it's better to conservative here.

bpintea commented on x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/TypeConverter.java:
```diff
@@ -581,28 +630,60 @@ private static OffsetDateTime asOffsetDateTime(Object val, EsType columnType, St
         throw new SQLFeatureNotSupportedException();
     }
 
-    private static byte safeToByte(long x) throws SQLException {
+    private static byte safeToByte(Number n) throws SQLException {
+        if (n instanceof BigInteger) {
+            try {
+                return ((BigInteger) n).byteValueExact();
+            } catch (ArithmeticException ae) {
+                throw new SQLException(format(Locale.ROOT, "Numeric %s out of range", n));
+            }
+        }
+        long x = n.longValue();
         if (x > Byte.MAX_VALUE || x < Byte.MIN_VALUE) {
-            throw new SQLException(format(Locale.ROOT, "Numeric %s out of range", Long.toString(x)));
+            throw new SQLException(format(Locale.ROOT, "Numeric %s out of range", n));
```
> for numbers it should be %d.

The argument to `%s` is a `Number` here, can be a fixed point, but can be a floating point too. `%d` [means](x-pack/plugin/sql/jdbc/src/main/java/org/elasticsearch/xpack/sql/jdbc/TypeConverter.java): `The result is formatted as a decimal integer`, which would result in an `IllegalFormatConversionException` for non-integers. 
Otoh `%s` would have `format` call `.toString()`, which should be safe for any Number.

costin replied:
👍 
Btw, `long.toString` behind the scenes calls `Long.toString()` - my point being that passing the number vs the static method call is more or less the same.

bpintea commented on x-pack/plugin/sql/qa/jdbc/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/JdbcTestUtils.java:
```diff
@@ -50,6 +64,36 @@ private JdbcTestUtils() {}
+        CLASS_TO_ES_TYPE = Collections.unmodifiableMap(aMap);
+    }
+
+    static EsType of(Class<?> clazz) {
```
> It's worth trying to do a basic backwards compatibility check between double and longs to reproduce the issue.

If an old client will sent a `"type": "double"` to a new server, the server will resolve the type by name, so no failure will occur. But if there's a specific failure path possible that I'm overseeing, let me know please.

Note: this particular `of(Class<?> clazz)` method is only used by tests.

bpintea commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/CsvTestUtils.java:
```diff
@@ -162,6 +162,8 @@ private static String resolveColumnType(String type) {
             case "sh":
                 return "short";
+            case "ul":
+                return "bigdecimal"; // CSV
```
The CSV JDBC driver [doesn't support BigInteger](https://github.com/simoc/csvjdbc/blob/2cafaea5d3a5d04ba5df47f11cc9eb4a6e96a080/src/main/java/org/relique/jdbc/csv/StringConverter.java#L646). I've extended the incomplete comment.

costin replied:
Shouldn't it be "biginteger" instead of "bigdecimal" though?
As a separate issue, it's considering whether the latest versions of JDBC CVS have added support for this and if not contributing this feature upstream.
It's a good opportunity to check whether an upgrade is warranted, I recall we looked at it in the past but can't remember the details.

bpintea committed:
Remove ShowColumns UL version-dep filtering

Unsigned long will now always show in 'SHOW COLUMNS', irrespective of
client's version (unlike 'SYS COLUMNS').

bpintea commented on x-pack/plugin/sql/qa/server/src/main/java/org/elasticsearch/xpack/sql/qa/jdbc/JdbcAssert.java:
```diff
@@ -261,6 +269,9 @@ private static void doAssertResultSetData(
                                 case "Int":
                                     columnClassName = "java.lang.Integer";
                                     break;
+                                case "BigDecimal":
```
This corresponds to the above CSV JDBC requirement.

bpintea commented:
@elasticmachine run elasticsearch-ci/part-2

bpintea committed:
Extract mapping reading out of loops

Extract the reading of an index mapping out of the loops in a couple of
tests.

costin reviewed (commented):
In this round of feedback:

1. Use the client version in the analyzer to change the server datatype from known to unknown. This removes any need to do filtering downstream to prevent the known type to leak to the client. This changes the versioning from being just a communication topic to resolving all server instances per client.
2. Mark the feature as experimental since clearly there are some issues with aggregations that are better addressed in ES - since this will take time, let's mark this accordingly and start the necessary discussions.
As an addition, for any new discovery, instead of providing a workaround, leave it as such or consider adding a validation rule.

For 1, please add a few tests where accessing the ul field, either in raw form or masked (such as a cast), blows up.

bpintea committed:
Resolve UL fields to unsupported in resolver

Resolve the unsigned long fiels to UNSUPPORTED directly in the index
resolver to remove the need for downstream type filtering.

Also rearange enum to prevent any possible serialization issues.

bpintea committed:
Merge branch 'master' into feat/unsigned_long

bpintea commented:
> Use the client version in the analyzer to change the server datatype from known to unknown.

The UL fields' datatype has been changed version dependent (in the IndexResolver, though, where the fields are created).

> Mark the feature as experimental

Sure, the docs PR that'll follow will note this support as experimental.

> please add a few tests where accessing the ul field

[Extended](https://github.com/elastic/elasticsearch/pull/65145/files#diff-6f95aacef2988567618b4250a265a6eadcebe2eb971653f7267a1a86c8e73c4fR1158) the [previous](https://github.com/elastic/elasticsearch/pull/65145/files#diff-dc85cc6d0343f1cb767a31671fca1577491c096c9e6b6f9fa59a7dbb97a2f024R326) test for this case.

bpintea committed:
Merge branch 'master' into feat/unsigned_long

bpintea commented:
@elasticmachine run elasticsearch-ci/part-1

Luegg commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/scalar/whitelist/InternalQlScriptUtils.java:
```diff
@@ -58,6 +59,10 @@ public static Number nullSafeCastNumeric(Number number, String typeName) {
         return number == null || Double.isNaN(number.doubleValue()) ? null : (Number) convert(number, fromTypeName(typeName));
     }
 
+    public static Number nullSafeCastToUnsignedLong(Number number) {
```
I think this new method creates a bwc issue in mixed node environments. Maybe consider bumping the `minCompatibleShardNode` version in `org/elasticsearch/xpack/sql/execution/search/Querier.java:147` and cover it in `SqlCompatIT`.

bpintea replied:
Changed the `minCompatibleShardNode`. 
The test with bwcVersion nodes would fail indeed, but due to missing UL SQL support, not just missing script, while script support current version nodes is validated in other tests, so not sure how to add a meaningful test specifically for this.

Luegg reviewed (commented):
A few more comments and one issue that might be worth addressing in this PR or in a follow up:

When multiplying intervals with ulongs, the result overflows instead of failing with an overflow error:

```
select ul, ul * interval 1 SECONDS from ulong

         ul         |  ul * interval 1 SECONDS  
--------------------+---------------------------
18446744073709551614|PT-2S                      
1                   |PT1S                       
9223372036854775808 |PT-2562047788015215H-30M-8S
```

bpintea committed:
Merge branch 'master' into feat/unsigned_long

bpintea committed:
Address review comments

Take SqlVersion -> Version out of a loop.
Set a minimum shard version at UL-version for searches.

elasticsearchmachine commented:
Hi @bpintea, I've created a changelog YAML for you.

Commit:
Update docs/changelog/65145.yaml

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/scalar/ScalarFunction.java:
```diff
@@ -150,7 +153,15 @@ protected ScriptTemplate scriptWithGrouping(GroupingFunction grouping) {
+        return field.dataType() != UNSIGNED_LONG
+            ? new ScriptTemplate(processScript(Scripts.DOC_VALUE), params, dataType())
+            : new ScriptTemplate(
+                processScript(formatTemplate(format("{ql}.", "nullSafeCastToUnsignedLong({})", Scripts.DOC_VALUE))),
```
Fixed, thanks.

bpintea commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -995,6 +1003,27 @@ private static void checkCastOnInexact(LogicalPlan p, Set<Failure> localFailures
 
+    private static void checkClientSupportsDataTypes(LogicalPlan p, Set<Failure> localFailures, SqlVersion version) {
+        p.output().forEach(e -> {
+            if (e.resolved() && isTypeSupportedInVersion(e.dataType(), Version.fromId(version.id)) == false) {
```
I've extracted that out of the loop, thanks.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java:
```diff
@@ -366,12 +369,16 @@ public void resolveAsMergedMapping(
         String indexWildcard,
         IndicesOptions indicesOptions,
         Map<String, Object> runtimeMappings,
+        Version version,
```
Why put the version in the IndexResolver and not the Analyzer?

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/Types.java:
```diff
@@ -15,32 +16,34 @@
+import static org.elasticsearch.xpack.ql.type.DataTypes.UNSIGNED_LONG;
 import static org.elasticsearch.xpack.ql.type.DataTypes.UNSUPPORTED;
 
 public abstract class Types {
```
The type discovery should be independent of the version - the compatibility layer needs to be resolved upstream.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java:
```diff
@@ -451,7 +452,7 @@ else if (plan instanceof OrderBy o) {
         private List<NamedExpression> filterUnsupportedProjections(List<NamedExpression> projections) {
             return projections.stream()
-                .filter(p -> p.resolved() == false || isTypeSupportedInVersion(p.dataType(), configuration.version()))
+                .filter(p -> p.resolved() == false || isTypeSupportedInVersion(p.dataType(), Version.fromId(configuration.version().id)))
```
Why not use directly `configuration.version()` instead of `Version.fromId(configuration.version().id)` ?

costin reviewed (commented):
Left yet another round.

bpintea committed:
Merge branch 'master' into feat/unsigned_long

bpintea committed:
Extract UL to Unsupported map transf. to own class

This extracts the unsigned_long replacement with UnsupportedEsField into
own class, `IndexCompatibility`, in case backwards compatibility needs
to apply. This leaves the responsability to ensure BWC to
IndexResolver's callers.

bpintea commented:
@elasticmachine run elasticsearch-ci/packaging-tests-unix-sample

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/session/EqlSession.java:
```diff
@@ -121,7 +120,6 @@ public void analyzedPlan(LogicalPlan parsed, ActionListener<LogicalPlan> listene
             indexWildcard,
             configuration.indicesOptions(),
             configuration.runtimeMappings(),
-            Version.CURRENT,
```
Instead of doing the filter manually, filter the IndexResult directly:
```
map(listener, r -> preAnalyzer.preAnalyze(parsed, IndexCompatibility.compatible(indexResolution.get(), Version.fromId(configuration.version().id));
```

This way the rules that call IndexResolution.get() will get the index already filtered - since I cannot think of an instance where that is needed, I think that's fine.

bpintea replied:
Maxes more sense indeed, fixed.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/session/SqlSession.java:
```diff
@@ -159,7 +158,7 @@ public void debugAnalyzedPlan(LogicalPlan parsed, ActionListener<RuleExecutor<Lo
                 includeFrozen,
                 configuration.runtimeMappings(),
-                Version.fromId(configuration.version().id),
+                // Version.fromId(configuration.version().id),
```
Leftover.

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/FieldAttributeTests.java:
```diff
@@ -346,59 +350,91 @@ public void testUnsignedLongVersionCompatibility() {
         for (SqlVersion version : List.of(preUnsignedLong, SqlVersion.fromId(INTRODUCING_UNSIGNED_LONG.id), postUnsignedLong)) {
             SqlConfiguration config = SqlTestUtils.randomConfiguration(version);
-            analyzer = new Analyzer(config, functionRegistry, getIndexResult, new Verifier(new Metrics()));
+            analyzer = new Analyzer(config, functionRegistry, loadMapping("mapping-numeric.json"), new Verifier(new Metrics()));
```
Why not pass `getIndexResult` and use mapping instead?

bpintea replied:
The mapping is mutated when making it "compatible" (on the first loop iteration), so it needs reloaded inside the loop; I've added a comment.

costin reviewed (commented):
Thanks for the changes - left minor polishing comments.

bpintea committed:
Move compat filtering out of the Analyzer

Move the compatibility filtering out of the Analyzer, which now receives
the IndexResolution already filtered; this now happens in the
SqlSession.

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/analyzer/FieldAttributeTests.java:
```diff
@@ -431,4 +464,22 @@ public void testExpandStarOnIndexWithoutColumns() {
         assertTrue(((Project) plan).projections().isEmpty());
     }
 
+    private static IndexResolution loadMapping(String mappingName) {
```
It helps to rename this method since it clashes with TypesTest.loadMapping and it's confusing to see the first returns a mapping while the latter an IndexResult.
How about `loadIndexResolution`?

bpintea replied:
Sure, will do (along with the needed `INTRODUCING_UNSIGNED_LONG` update after the pending version bump).

costin reviewed (approved):
LGTM from my side.

astefan reviewed (approved):
LGTM

Luegg reviewed (approved):
Lgtm 🚀

bpintea committed:
Merge branch 'master' into feat/unsigned_long

bpintea committed:
Apply review comments. Bump UL intro version

Rename loadMapping to loadIndexResolution.
Bump INTRODUCING_UNSIGNED_LONG version.

bpintea commented:
@elasticmachine run elasticsearch-ci/docs

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feat/unsigned_long

jrodewig commented:
@elasticmachine run elasticsearch-ci/docs
